exiftool.go

```diff
@@ -4,53 +4,97 @@ import (
 	"bufio"
 	"bytes"
 	"encoding/json"
-	"errors"
 	"fmt"
 	"io"
 	"os"
 	"os/exec"
 	"sync"
-)
 
-var binary = "exiftool"
-var executeArg = "-execute"
-var initArgs = []string{"-stay_open", "True", "-@", "-", "-common_args"}
-var extractArgs = []string{"-j"}
-var closeArgs = []string{"-stay_open", "False", executeArg}
-var readyToken = []byte("{ready}\n")
-var readyTokenLen = len(readyToken)
+	"errors"
+)
 
 // ErrNotExist is a sentinel error for non existing file
 var ErrNotExist = errors.New("file does not exist")
 
+// config contains the configuration used by the module
+type config struct {
+	binary        string
+	executeArg    string
+	initArgs      []string
+	extractArgs   []string
+	closeArgs     []string
+	readyToken    []byte
+	readyTokenLen int
+}
+
+// NewExiftoolConfig creates a new default configuration to be used with Exiftool. Pass configuration functions
+// to change the default configuration
+func NewExiftoolConfig(opts ...func(*config) error) (*config, error) {
+	config := config{
+		binary:        "exiftool",
+		executeArg:    "-execute",
+		initArgs:      []string{"-stay_open", "True", "-@", "-", "-common_args"},
+		extractArgs:   []string{"-j"},
+		closeArgs:     nil,
+		readyToken:    []byte("{ready}\n"),
+		readyTokenLen: 0,
+	}
+
+	config.closeArgs = []string{"-stay_open", "False", config.executeArg}
+
+	for _, opt := range opts {
+		if err := opt(&config); err != nil {
+			return nil, fmt.Errorf("error setting configuration options: %w", err)
+		}
+	}
+
+	config.readyTokenLen = len(config.readyToken)
+
+	return &config, nil
+}
+
 // Exiftool is the exiftool utility wrapper
 type Exiftool struct {
+	config  *config
```

Comment:
What's the interest of having a `config` field in the `Exiftool` struct ? Furthermore if its type is a pointer (`*config`) : what's the behaviour if it's nil ?
I honestly think that such a `config` struct is not useful inside the `Exiftool` struct : it would have been if it could change between `Exiftool` isntance, but there, it does not.
It is a memory loss.
If you create 10 `Exiftool` there will be 10 `config` that are identical.
I personnaly prefer using global variables (unfortunately, they can't be constants since slices can't be constants).

Reply:
It's possible to create a config and use with several Exiftool instances, by creating one using the NewExiftoolconfig function, and by passing a configuration function to NewExiftool that sets *config to point to the prevoiusly created config. It's also possible to create a check to verify *config is not nil before using it, but as the code is now it will always be created with the default params, unless a user manually overrides it with option functions in their code. But in the end, if you prefer the other approach then that is of course fine! :)

Reply:
I do prefer the "constant" versions.
It seems more straightforward to me and the code seems more readable.

Reply:
@Hexoplon, do you plan to make the last changes and come back to constants or do I integrate myself manually all the other great changes that you proposed and that remains ?

Reply:
I'll make the last changes!